### PR TITLE
feat: provide import completion for include and images

### PR DESCRIPTION
**Pull request by d-koppenhagen**
related to #377

### CODE REVIEW

**Comment by d-koppenhagen** [comment]
still WIP

**Comment by d-koppenhagen** [comment]
`editor.autoClosingBrackets` makes some problems I need to solve: https://stackoverflow.com/questions/65079743/prevent-autoclosingbrackets-when-inserting-content-as-result-of-registercompleti

File: src/providers/asciidoc.provider.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    kind: vscode.CompletionItemKind.Folder,
+    sortText: '..',
+  }
+  const globalVariableDefinitions = documentText.match(/:\S+:.*/g)
+
+  let variablePathSubstitutions = []
+  // TODO: prevent editor.autoClosingBrackets at this point until finished inserting
+  const editorConfig = vscode.workspace.getConfiguration('editor');
+  const doAutoCloseBrackets = editorConfig.get('autoClosingBrackets') === 'always';
+  if (globalVariableDefinitions) {
```
Comment (danyill) [inline src/providers/asciidoc.provider.ts:76]: We already have autocompletion on the brace character `{` from the Asciidoc AST (see https://github.com/asciidoctor/asciidoctor-vscode/blob/master/src/features/attributeCompleter.ts).

File: src/providers/asciidoc.provider.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+/**
+ * Provide Completion Items
+ */
+async function provide(
+  context: Context
+): Promise<vscode.CompletionItem[]> {
+  const documentText = context.document.getText();
+  const pathExtractedFromIncludeString = context.textFullLine.replace('include::', '').replace('image::', '').replace('image:', '');
```
Comment (danyill) [inline src/providers/asciidoc.provider.ts:42]: Ideally we would resolve `image::` references in the completion relative to the document base directory and the `imagesdir` if set.

**Comment by danyill** [comment]
Thanks for this great contribution, I've left a few comments if you're interested in examining a bit further a few small aspects but otherwise it looks good to me.

File: src/providers/asciidoc.provider.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+  return [
+    levelUpCompletionItem,
+    ...variablePathSubstitutions,
+    ...items.map((child) => {
+      const result = createPathCompletionItem(child);
+      result.insertText = result.kind === vscode.CompletionItemKind.File ? child.file + '[]' : child.file
+      return result;
+    }),
+  ];
```
[the remark this answers is not in the document]
Comment (d-koppenhagen) [reply src/providers/asciidoc.provider.ts]: solved here 
ee69fe2

**Comment by joaompinto** [comment]
Thanks for the contribution, looks good to me.

**Comment by joaompinto** [comment]
I have tested the include:: autocomplete, it does not provide me a list of the files, screensot:
autocomplete_issue

Did I miss something ?

**Comment by d-koppenhagen** [comment]
Not sure but when I check out the master branch now and run the extension locally it seems to work fine.
Did you miss to scroll down? probably the suggestion order need to be re-thinked but I am not sure what will appear first in the list as other extension may contribute also to the list.

Here is a screenshot of my local test workspace:

Screenshot 2021-01-23 at 13 10 30

**Comment by eiswind** [comment]
I found that in 2.9.5 it doesn't work when I have attributes in the include path.
include::{srcdir}/002_config/src/main/java/de/eiswind/training/spring/config/Application.java

another thing I noticed is that the base path is the project directory. Shouldn't inlcudes be relative to the adoc file?

Any help would be apprieciated.

**Pull request merged**